engine-spring/core/src/test/java/org/camunda/bpm/engine/spring/test/scripttask/ScriptTaskTest.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership. Camunda licenses this file to you under the Apache License,
+ * Version 2.0; you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.bpm.engine.spring.test.scripttask;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.camunda.bpm.engine.RepositoryService;
+import org.camunda.bpm.engine.RuntimeService;
+import org.camunda.bpm.engine.repository.Deployment;
+import org.camunda.bpm.engine.runtime.ProcessInstance;
+import org.camunda.bpm.engine.spring.test.SpringProcessEngineTestCase;
+import org.camunda.bpm.model.bpmn.Bpmn;
+import org.camunda.bpm.model.bpmn.BpmnModelInstance;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+/**
+ *
+ * @author Laszlo Palossy
+ *
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(locations = {
+    "classpath:org/camunda/bpm/engine/spring/test/scripttask/ScriptTaskTest-applicationContext.xml" })
+public class ScriptTaskTest extends SpringProcessEngineTestCase {
+
+  @Autowired
+  private RuntimeService runtimeService;
+
+  @Autowired
+  private RepositoryService repositoryService;
+
+  @Autowired
+  Testbean testbean;
+
+  private static final String JAVASCRIPT = "javascript";
+  private static final String PYTHON = "python";
+  private static final String GROOVY = "groovy";
+  private static final String JUEL = "juel";
+
+  private List<String> deploymentIds = new ArrayList<String>();
+
+  @After
+  public void after() {
+    for (String deploymentId : deploymentIds) {
+      repositoryService.deleteDeployment(deploymentId, true);
+    }
+  }
+
+  @Test
+  public void testJavascriptProcessVarVisibility() {
```

Comment:
You can reduce the code of the test by Parameterizing the class, like in `org.camunda.bpm.engine.test.api.variables.PrimitiveTypeValueSerializationTest`.

Reply:
All chaned source classes were formatted like in the style guide (with no `@Author` tags, and other formatting rules).

I have made two new versions for `ScriptTaskTest`:
1. with optimized code, no code duplications (is committed now)
2. runs with `@RunWith(Parameterized.class)` (the source is attached)

The (1) version is committed now, because it runs much faster (~8 sec vs. ~16 sec).
The source code of the test class really did not look pretty :).
Now looks like this:
```java
  @Test
  public void testSpringBeanVisibility() {
    testSpringBeanVisibility(JAVASCRIPT, "execution.setVariable('foo', testbean.name);");
    testSpringBeanVisibility(GROOVY, "execution.setVariable('foo', testbean.name)\n");
    testSpringBeanVisibility(PYTHON, "execution.setVariable('foo', testbean.name)\n");
    testSpringBeanVisibility(JUEL, "${execution.setVariable('foo', testbean.name)}");
  }
```
It just calls a common method for each scriptformat now.

__Some remarks for the (2.) `@RunWith(Parameterized.class)` version:__

Please take a look at it, the source is attached!
The time is lost at initialzation.
The sample `PrimitiveTypeValueSerializationTest` initializes so:
```java
  @Rule
  public ProcessEngineRule rule = new ProvidedProcessEngineRule();

  @Before
  public void setup() {
    runtimeService = rule.getRuntimeService();
    repositoryService = rule.getRepositoryService();
    ....
  }
```
I could make it only so:
```java
  @Before
  public void setup() {
    applicationContext = new ClassPathXmlApplicationContext(
        "org/camunda/bpm/engine/spring/test/scripttask/ScriptTaskTest-applicationContext.xml");

    runtimeService = applicationContext.getBean("runtimeService", RuntimeService.class);
    repositoryService = applicationContext.getBean("repositoryService", RepositoryService.class);
  }
```
I coud not use the `ProvidedProcessEngineRule` class, as it is in the *src/test/java* folder of the *camunda-engine* project, so it is not included in the dependency jar.
(With adjusting the pom's of the two projects it could be used, if needed.)

Simple autowring `runtimeService` and `repositoryService` with `@ContextConfiguration` did not work either, they were null, i think initialization does not happen in the right order for us.

What do yo think, should we rather use the `@RunWith(Parameterized.class)` version? Perhaps with adjusting the pom's to get `ProvidedProcessEngineRule`?
Or just let the committed version of the test class?

Reply:
Attaching does not work for me...... So here is the code of the mentioned file:

```java
package org.camunda.bpm.engine.spring.test.scripttask;

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

import org.camunda.bpm.engine.RepositoryService;
import org.camunda.bpm.engine.RuntimeService;
import org.camunda.bpm.engine.repository.Deployment;
import org.camunda.bpm.engine.runtime.ProcessInstance;
import org.camunda.bpm.model.bpmn.Bpmn;
import org.camunda.bpm.model.bpmn.BpmnModelInstance;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameter;
import org.junit.runners.Parameterized.Parameters;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.context.support.ClassPathXmlApplicationContext;

@RunWith(Parameterized.class)
public class ScriptTaskTestParametrized {

  @Parameters
  public static Collection<Object[]> scriptAttributes() {
    return Arrays.asList(new Object[][] { //
        { JAVASCRIPT, "execution.setVariable('foo', testbean.name);" }, //
        { PYTHON, "execution.setVariable('foo', testbean.name)\n" }, //
        { GROOVY, "execution.setVariable('foo', testbean.name)\n" }, //
        { JUEL, "${execution.setVariable('foo', testbean.name)}" } //
    });
  }

  @Parameter(0)
  public String scriptFormat;

  @Parameter(1)
  public String scriptText;

  protected static final String JAVASCRIPT = "javascript";
  protected static final String PYTHON = "python";
  protected static final String GROOVY = "groovy";
  protected static final String JUEL = "juel";
  protected static final String TEST_BEAN_NAME = "name property of testbean";

  protected ApplicationContext applicationContext;
  protected RuntimeService runtimeService;
  protected RepositoryService repositoryService;

  protected List<String> deploymentIds = new ArrayList<String>();

  @Before
  public void setup() {
    applicationContext = new ClassPathXmlApplicationContext(
        "org/camunda/bpm/engine/spring/test/scripttask/ScriptTaskTest-applicationContext.xml");

    runtimeService = applicationContext.getBean("runtimeService", RuntimeService.class);
    repositoryService = applicationContext.getBean("repositoryService", RepositoryService.class);
  }

  @After
  public void after() {
    for (String deploymentId : deploymentIds) {
      repositoryService.deleteDeployment(deploymentId, true);
    }
    ((ConfigurableApplicationContext) applicationContext).close();
  }

  /**
   * Test if a Spring bean is visible for scripting for the scripftormat
   * 
   * @param scriptFormat
   *          the scriptformat like 'javascript', 'groovy', etc.
   * @param scriptText
   *          sets execution variable 'foo' to the testbean's name property
   */
  @Test
  public void testSpringBeanVisibility() {

    deployProcess(scriptFormat, scriptText);

    ProcessInstance pi = runtimeService.startProcessInstanceByKey("testProcess");

    Object variableValue = runtimeService.getVariable(pi.getId(), "foo");
    assertEquals(TEST_BEAN_NAME, variableValue);
  }

  protected void deployProcess(BpmnModelInstance process) {
    Deployment deployment = repositoryService.createDeployment().addModelInstance("testProcess.bpmn", process).deploy();
    deploymentIds.add(deployment.getId());
  }

  protected void deployProcess(String scriptFormat, String scriptText) {
    BpmnModelInstance process = createProcess(scriptFormat, scriptText);
    deployProcess(process);
  }

  protected BpmnModelInstance createProcess(String scriptFormat, String scriptText) {

    return Bpmn.createExecutableProcess("testProcess").startEvent().scriptTask().scriptFormat(scriptFormat)
        .scriptText(scriptText).userTask().endEvent().done();
  }
}

```

Reply:
Hi @palossyl ,

Thanks for the thorough analysis. Since parameterizing the test makes it a lot slower, I'm okay with using the committed version. However, it this case it would be better to create 4 separate tests for each scripting engine. The way it's implemented now, it would be harder for debugging if a single script engine fails.

Best,
Nikola

Reply:
Hi Nikola!

There are 4 separate test now for each scripting engine!

Best,
          Laszlo